src/framework/ui/internal/navigationcontroller.cpp

```diff
@@ -1198,6 +1199,40 @@ void NavigationController::doTriggerControl()
     activeCtrl->trigger();
 }
 
+void NavigationController::onConfirm()
+{
+    TRACEFUNC;
+    MYLOG() << "====";
+    INavigation::EventPtr e = Event::make(Event::Confirm);
+
+    INavigationControl* activeCtrl = this->activeControl();
+    if (activeCtrl) {
+        activeCtrl->onEvent(e);
+        if (e->accepted) {
+            return;
+        }
+    }
+
+    INavigationPanel* activePanel = this->activePanel();
+    if (activePanel) {
+        activePanel->onEvent(e);
+        if (e->accepted) {
+            return;
+        }
+    }
+
+    INavigationSection* activeSec = this->activeSection();
+    if (activeSec) {
+        activeSec->onEvent(e);
+        if (e->accepted) {
+            return;
+        }
+    }
```

Comment:
The following implementation would avoid some double work being done by calling `activePanel` and `activeSection` after `activeControl`
```c++
    INavigationSection* activeSec = activeSection();
    if (activeSec) {
        INavigationPanel* activePanel = findActive(activeSec->panels());
        
        if (activePanel) {
            INavigationControl* activeCtrl = findActive(activePanel->controls());
            
            if (activeCtrl) {
                activeCtrl->onEvent(e);
                if (e->accepted) {
                    return;
                }
            }

            activePanel->onEvent(e);
            if (e->accepted) {
                return;
            }
        }

        activeSec->onEvent(e);
        if (e->accepted) {
            return;
        }
    }
```
But the nested `if`s are maybe a disadvantage; I'll let @Eism decide.

Reply:
Or

```
    INavigationSection* activeSec = activeSection();
    INavigationPanel* activePanel = activeSec ? findActive(activeSec->panels()) : nullptr;
    INavigationControl* activeCtrl = activePanel ? findActive(activePanel->controls()) : nullptr;

    if (activeCtrl) {
        activeCtrl->onEvent(e);
        if (e->accepted) {
            return;
        }
    }

    if (activePanel) {
        activePanel->onEvent(e);
        if (e->accepted) {
            return;
        }
    }

    if (activeSec) {
        activeSec->onEvent(e);
        if (e->accepted) {
            return;
        }
    }
```